go/vt/schema/online_ddl.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+
+	"vitess.io/vitess/go/vt/topo"
+)
+
+var (
+	migrationBasePath = "schema-migration"
+)
+
+// MigrationBasePath is the root for all schema migration entries
+func MigrationBasePath() string {
+	return migrationBasePath
+}
+
+// MigrationRequestsPath is the base path for all newly received schema migration requests.
+// such requests need to be investigates/reviewed, and to be assigned to all shards
+func MigrationRequestsPath() string {
+	return fmt.Sprintf("%s/requests", MigrationBasePath())
+}
+
+// MigrationReviewedPath is the base path for schema migrations that have been reviewed. Kept for
+// historical reference
+func MigrationReviewedPath() string {
+	return fmt.Sprintf("%s/reviewed", MigrationBasePath())
+}
+
+// MigrationJobsKeyspacePath is the base path for a tablet job, by keyspace
+func MigrationJobsKeyspacePath(keyspace string) string {
+	return fmt.Sprintf("%s/jobs/%s", MigrationBasePath(), keyspace)
+}
+
+// MigrationJobsKeyspaceShardPath is the base path for a tablet job, by keyspace and shard
+func MigrationJobsKeyspaceShardPath(keyspace, shard string) string {
+	return fmt.Sprintf("%s/%s", MigrationJobsKeyspacePath(keyspace), shard)
+}
+
+// OnlineDDLStatus is an indicator to a online DDL status
+type OnlineDDLStatus string
+
+const (
+	OnlineDDLStatusRequested OnlineDDLStatus = "requested"
+	OnlineDDLStatusReviewed  OnlineDDLStatus = "reviewed"
+	OnlineDDLStatusCancelled OnlineDDLStatus = "cancelled"
+	OnlineDDLStatusQueued    OnlineDDLStatus = "queued"
+	OnlineDDLStatusReady     OnlineDDLStatus = "ready"
+	OnlineDDLStatusRunning   OnlineDDLStatus = "running"
+	OnlineDDLStatusComplete  OnlineDDLStatus = "complete"
+	OnlineDDLStatusFailed    OnlineDDLStatus = "failed"
+)
```

Comment:
`gh-ost` supports [hooks](https://github.com/github/gh-ost/blob/master/doc/hooks.md), which we utilize. One of those hooks is `on-status`, which fires every `1minute`. You can use that as a liveness indicator. 
Back when I worked on [skeefree](https://github.blog/2020-02-14-automating-mysql-schema-migrations-with-github-actions-and-more/), I used it as a liveness/health indicator, such that if I haven't seen a report in past `10minutes`, I assumed the migration to be dead.

In our current (still evolving) setup, things are somewhat simpler, because I know for fact that `gh-ost` runs on the master tablet's host, and so can further ask the tablet to communicate with it. That is, we will still use `on-status`, but then again, the tablet may also check that the `gh-ost` process is running (by communicating with the OS), or forcibly `kill -9` it, etc.